library/core/src/option.rs

```diff
@@ -1750,7 +1751,11 @@ impl<T> Option<Option<T>> {
     /// ```
     #[inline]
     #[stable(feature = "option_flattening", since = "1.40.0")]
-    pub fn flatten(self) -> Option<T> {
-        self.and_then(convert::identity)
+    #[rustc_const_unstable(feature = "const_option", issue = "67441")]
+    pub const fn flatten(self) -> Option<T> {
+        match self {
+            Some(inner) => inner,
+            None => None,
+        }
```

Comment:
Does it make sense to also make `and_then` unstably const so you don't have to change the body?

Reply:
`and_then` takes a function as an argument, which afaik cannot currently be declared `const`. The RFC for const trait impls hasn't even been merged yet, surprisingly.